### ISSUE TITLE
Fresh Barrier Install Shows is:openERROR: ssl certificate doesn't exist: /home/thomas/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem

### ISSUE THREAD

**Issue by ghost** [input-leap/input-leap#231]
> This issue has been migrated from old Barrier Github repository https://github.com/debauchee/barrier/issues/231
> 
> Issue created on: 2019-01-18 by @TafThorne
> Issue last updated on: 2021-11-08

### Operating Systems ###

Server: Ubuntu 18.04

Client: Microsfot Windows 10 Version 1803 (OS Build 17134.523)

### Barrier Version ###

Server: 2.2.0-snapshot-53ebc47a

Client: 2.1.0-RELEASE-0b2dfd80

### Steps to reproduce bug ###

1. Go to https://github.com/debauchee/barrier/wiki to Download and install the Windows binary.
2. Go to https://github.com/debauchee/barrier/releases to install the latest release, which for Ubuntu is a Flatpak
3. Start the client on the Windows machine, set to client mode and give server IP address.
4. Start the server on the Ubuntu machine with `flatpak run com.github.debauchee.barrier`
5. Click the "Configure Sever..." button and enter the client details.
6. Start Barrier
7. Observe errors in the Barrier > Show Log

```
[2019-01-18T11:41:57] INFO: OpenSSL 1.1.1  11 Sep 2018
[2019-01-18T11:41:57] ERROR: ssl certificate doesn't exist: /home/thomas/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem
[2019-01-18T11:42:13] INFO: OpenSSL 1.1.1  11 Sep 2018
[2019-01-18T11:42:13] ERROR: ssl certificate doesn't exist: /home/thomas/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem
[2019-01-18T11:42:29] INFO: OpenSSL 1.1.1  11 Sep 2018
```

### Other info ###

* When did the problem start to occur? On a fresh install of what I (probaby mistakenly) thought was the latest releases.  
* Is there a way to work around it? IDK, probably trying to get the same version on both ends?
* Does this bug prevent you from using Barrier entirely? Yes

---

> Commented on: 2019-01-18 by @TafThorne

I can confirm that the file does not exist:

```
$ ls ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/
$ locate Barrier.pem
```

---

> Commented on: 2019-01-18 by @TafThorne

I thought this could be similar to #142 but allowing barrier.exe and barriers.exe through the firewall did not change the message in the logs.

#'171 also looked related but I cannot work out where the Barrier.pem shown there comes from.  Following the supplied link did not really change anything. 

I can at least confirm this is related to having the SSL config enabled.  With that disabled I am able to get the client and server to work together.

---

> Commented on: 2019-01-18 by @AdrianKoshka

That's interesting, I'm a daily user of the flatpak and barrier managed to generate an SSL certificate for me just fine.

```
alc@am1m-s2h ~ % ls -l .var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem 
-rw-rw-r--. 1 alc alc 1649 May 21  2018 .var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem
```

---

> Commented on: 2019-01-21 by @eurythmia

As a workaround, I used the synergy instructions from this page: https://wiki.archlinux.org/index.php/synergy#Set_up_encryption_on_server

**TL;DR:**
```
mkdir -p ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Fingerprints
openssl req -x509 -nodes -days 365 -subj /CN=Barrier-newkey rsa:4096-keyout ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem -out ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem
openssl x509 -fingerprint -sha2 -noout -in ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem > ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Fingerprints/Local.txt
```
Then enable encryption from the settings on both the client and server, restart both, and accept the new server certificate from your client ... should be good to go from there.

Edit:
change keylength from `1024` to `4096`
change `-sha1` to `-sha2`

---

> Commented on: 2019-01-21 by @AdrianKoshka

I'll leave this issue open just in case someone else also has this issue. Also you might want to bump that key-length from `1024` to `4096`, and use `sha2` instead of `sha1`.

---

> Commented on: 2019-01-21 by @eurythmia

fair point. I usually use the 4K keys ... I just got lazy this time around and copy/pasted. Comment updated with @AdrianKoshka 's suggestions.

Edit: punctuation.

---

> Commented on: 2019-01-22 by @TafThorne

When I run the `openssl req` command I get an error message:

```
$ openssl req -x509 -nodes -days 365 -subj /CN=Barrier-newkey rsa:4096-keyout ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem -out ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem
req: Use -help for summary.
```
It looks like you have lost a couple of spaces from the command.  The following works:

```
$ openssl req -x509 -nodes -days 365 -subj /CN=Barrier -newkey rsa:4096 -keyout ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem -out ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem
Generating a 4096 bit RSA private key
......................................................................................................................................................++
.......................................................................................................................................................................++
writing new private key to '/home/thomas/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem'
-----
```

The `openssl x509` command then also fails.
```
$ openssl x509 -fingerprint -sha2 -noout -in ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem > ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Fingerprints/Local.txt
x509: Unknown digest sha2
x509: Use -help for summary.
```

I believe that the workaround advice should be updated to:

```
mkdir -p ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Fingerprints
openssl req -x509 -nodes -days 365 -subj /CN=Barrier -newkey rsa:4096 -keyout ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem -out ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem
openssl x509 -fingerprint -sha1 -noout -in ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Barrier.pem > ~/.var/app/com.github.debauchee.barrier/data/barrier/SSL/Fingerprints/Local.txt
```

At least in my case.

---

> Commented on: 2019-01-22 by @TafThorne

Now does the absence of sha2 support explain the cause of my issue?  If whatever is expected to automatically generate the keys is trying to use sha2 and silently failing.  

I have version 1.1.0g of OpenSSL.
```
$ openssl version
OpenSSL 1.1.0g  2 Nov 2017
```
What version is expected to have the sha2 support?  It seems that I can use either `-sha256` or `-sha512` in the command line without getting an error.

---

> Commented on: 2019-01-22 by @TafThorne

I am pleased to report that with the workaround in place (using sha512 in this case) I have been able to connect the Windows client to the flatpak version of barrier.  

Thank you for your help in getting me up and running.  Please let me know if there is anything more I can do to help find the cause of the issue and possible solution.

---

> Commented on: 2019-02-12 by @julius59

Same setup and same behaviour as in the bug description here.

---

> Commented on: 2019-03-29 by @nedart

I had this same issue. Both client and server were version 2.2.0-snapshot-53ebc47a. Using the commands in the post by @TafThorne did the trick.

---

> Commented on: 2019-05-30 by @tassosblackg

I had the same issue, it worked for me too!

---

> Commented on: 2019-06-14 by @hatzkel

This just crept up on me out of the blue after this past update. Had to follow the instructions above as well.

---

> Commented on: 2019-06-15 by @obasilakis

Hey, I tried @TafThorne 's commands and I'm getting this:

` [2019-06-15T17:20:12] INFO: OpenSSL 1.1.1b  26 Feb 2019`
` [2019-06-15T17:20:12] ERROR: could not use ssl certificate`
` [2019-06-15T17:20:12] ERROR: error:0909006C:PEM routines:get_name:no start line`

anyone knows what's the issue?

---

> Commented on: 2019-07-17 by @hifi

Not trying to sound rude but why this hasn't been fixed in the official Flatpak build as it seems to affect all new installs, including me today and is a real _barrier_ to entry for new users?

---

> Commented on: 2019-07-24 by @AdrianKoshka

So, as was probably obvious, the `openssl` command-line tools aren't shipped, which prevents the certificate from being generated, I'm working on fixing this finally. I'm sorry.

---

> Commented on: 2019-07-24 by @AdrianKoshka

Truly sorry this was an issue for so long, it really shouldn't have been. After work it can be hard to find the energy to work on OSS. 

---

> Commented on: 2019-07-24 by @hifi

Indeed, thank you for fixing the issue.

My point was that the Flatpak package is likely one of the main channels to install Barrier and since it was effectively broken for all new users it was a bit frustrating to find the issue had been open for so long.

---

> Commented on: 2019-07-24 by @AdrianKoshka

Fixed for me on a personal machine at work, before I had no cert.

![Screenshot_2019-07-24_10-19-59](https://user-images.githubusercontent.com/8690255/61801819-71421c80-adfd-11e9-99a9-fde2758f525a.png)

---

> Commented on: 2019-10-10 by @AdrianKoshka

`openssl` is shipped with barriers flatpak, and I believe this has been fixed.

---

> Commented on: 2020-07-16 by @fbidu

Hey people, I just ran into this issue on a fresh install using `snapd` on fedora!

```
mkdir -p /home/fbidu/snap/barrier-kvm/2/.local/share/barrier/SSL/ 
cd /home/fbidu/snap/barrier-kvm/2/.local/share/barrier/
openssl req -x509 -nodes -days 365 -subj /CN=Barrier -newkey rsa:4096 -keyout Barrier.pem -out Barrier.pem
```

Worked for me!

---

> Commented on: 2021-11-03 by @4F2E4A2E

This is how I've got it running on Windows:

- Install git for windows
- cd into `'C:\Users\<user>\AppData\Local\Barrier\SSL\'`
- run in the cmd `openssl req -x509 -nodes -days 365 -subj /CN=Barrier -newkey rsa:4096 -keyout Barrier.pem -out Barrier.pem`
- restart barrier

---

> Commented on: 2021-11-06 by @cfarvidson

I had the exact same issue on macOS Monterey (12.0.1) today.

Solved it by running the openssl command described in @4F2E4A2E post.

- cd into `/Users/<user>/Library/Application Support/barrier/SSL`
- run in the cmd `openssl req -x509 -nodes -days 365 -subj /CN=Barrier -newkey rsa:4096 -keyout Barrier.pem -out Barrier.pem`
- restart barrier

---

> Commented on: 2021-11-06 by @shymega

This should be opened again, as it appears to be an _ongoing_ issue with Barrier. Reopening.

---

> Commented on: 2021-11-08 by @CaptainKraft

I also had this problem with a fresh install on Arch Linux. Also, the workaround did not work for me due to differing paths. Here are the commands I had to use to get this to work:

```
mkdir -p ~/.local/share/barrier/SSL/Fingerprints
openssl req -x509 -nodes -days 365 -subj /CN=Barrier -newkey rsa:4096 -keyout ~/.local/share/barrier/SSL/Barrier.pem -out ~/.local/share/barrier/SSL/Barrier.pem
openssl x509 -fingerprint -sha1 -noout -in ~/.local/share/barrier/SSL/Barrier.pem > ~/.local/share/barrier/SSL/Fingerprints/Local.txt
```

**Comment by claudiusraphael** [comment]
I just commented on a related issue in the barrier repo, which i couldn't find here, but might be noteworthy to post here as well.

Original issue 1609: [Barrier.pem not generated on MacOS install for v2.4.0](https://github.com/debauchee/barrier/issues/1609#top)

Copy of my comment over there:

> Oh, I made it work between Windows and Mac M1 with both machines as Client and Server. The important point, missing in docs, here is that to use a device as the Server, it must have the SSL fingerprint generated and this software for some reason doesn't do it automatically in Windows or Mac either. Hence, you see SSL disabled in it. And, if the SSL is disabled in the Server, the client won't connect to it. You can possibly (I've not checked it) disable this behavior by going in the software settings Barrier -> Change Settings and Disable SSL but, for someone like me who wanted to do this with SSL, here's the solution.
> 
> So, to use it a machine as a server, you've to generate SSL certificate manually.
> 
> To do that in MAC if you're willing to use MAC as the server:
> 
> Open a terminal to its SSL location as in Library -> Application Support -> barrier -> SSL. You can use finder or the command
> //Bypassing Space using backward Slash \
> cd /Users/USERNAME/Library/Application\ Support/barrier/SSL 
> Generate the SSL certificate
> openssl req -new -x509 -sha256 -days 999 -nodes -out Barrier.pem -keyout Barrier.pem
> Restart the App in MAC and you'll see the generated SSL fingerprint.
> Ready to use it as a server.
> To generate SSL in Windows for using it as a Server.
> 
> Open CMD/PowerShell to its SSL location as in C:\Users\USERNAME\AppData\Local\Barrier\SSL. You can use the explore or the command
> cd C:\Users\USERNAME\AppData\Local\Barrier\SSL
> Generate the SSL certificate
> openssl req -x509 -nodes -days 365 -subj //CN=Barrier -newkey rsa:4096 -keyout Barrier.pem -out Barrier.pem
> Restart the App in Windows and you'll see the generated SSL fingerprint.
> Ready to use it as a server.
> By default, the client's SSL is disabled but if you want to go full SSL, you can enable it in Barrier Menu -> Change Settings -> Require Client Certificate.
> 
> This is what the docs/Reame.md should mention and it will resolve possibly all issues mentioning it's stuck in MAC or doesn't work at all.
> 
> After making it work, I understand why I found a few comments about this software even before finding this that mentioned it would be awesome if it worked.
> 
> Thank you the community support, I've switched from Share mouse to this and I like it here.
> 
> Edit:
> If you don't care about the SSL or don't want to go the aforementioned technical way, you can just disable it in the Settings menu as mentioned above in the first para and it will work.

Luckily found this comment after spending hours to somehow figure out why on earth it is not possible to establish a connection between two MacBook Pro running Ventura.

This information is a must to include in the Readme or at least Wiki.

On macos Ventura only if you right from the start select client a ssl-key will be generated, if you choose server from the get go it will not be created, even if you remove the application and the .plist in users Libraries for Barrier later on (for example to overcome the problem that barrier can not auto-config after a hostname-change) - why, i was not able to find out.

Also, it is not possible to deactivate SSL on the client, because without ever having had a connection between server and client established you could not access the settings (which can not be accessed at all btw. as there simply is no such entry visible in macos Ventura and barrier 2.4), so creating the SSL-key manually becomes a must for the server.

Why SSL is by default activated and that default can't be overridden on setup is another riddle, especially in a trusted private local network, which i would assume is the most often given case for users of barrier.

Regarding the creation of the SSL-key it is worth noting that the 2-letter country-code that is asked for first is a necessity, while all other requested information can be filled blank by entering a dot on each question.

Another hurdle on macos is that if you do try to delete/reinstall barrier it will not let you proceed because it does not actually check if "Accessibility" right is enabled in Privacy & Security for barrier, meaning the entry will still be there after removing barrier and still enabled but barrier does not detect that, hence not able to go on. Just toggling it to disable quit barrier, launch barrier then enable the entry also has no effect - it is necessary to remove the entry and manually add it back in, as barrier does not add the entry if missing if it was present before.

Leaving those notes for those that by accident or sheer luck stumble over this or for those that care to summarize it and add a note to the wiki/readme for first-time setup on macos.

Good luck.

**Comment by sithlord48** [comment]
this has been fixed at some point unsure when as fresh install create certs correctly now

**Comment by shymega** [comment]
Resolved.
